# PART 2

In [1]:
import torch.nn as nn
import torch.nn.functional as F
from torchsummary import summary
import torch

from tqdm import tqdm

In [2]:
from __future__ import print_function
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms

In [3]:
use_cuda = torch.cuda.is_available()
device = torch.device("cuda" if use_cuda else "cpu")

In [4]:
device

device(type='cuda')

In [5]:
torch.manual_seed(1) #setting manual seed
batch_size = 128 #selecting a batch size of 128

kwargs = {'num_workers': 1, 'pin_memory': True} if use_cuda else {}

#defining dataloader for train
train_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=True, download=True,
                    transform=transforms.Compose([
                        transforms.ToTensor(),
                        transforms.Normalize((0.1307,), (0.3081,))
                    ])),
    batch_size=batch_size, shuffle=True, **kwargs)
#defining dataloader for train
test_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=False, transform=transforms.Compose([
                        transforms.ToTensor(),
                        transforms.Normalize((0.1307,), (0.3081,))
                    ])),
    batch_size=batch_size, shuffle=True, **kwargs)

100%|██████████| 9912422/9912422 [00:00<00:00, 102601609.63it/s]


Extracting ../data/MNIST/raw/train-images-idx3-ubyte.gz to ../data/MNIST/raw



100%|██████████| 28881/28881 [00:00<00:00, 38467987.88it/s]


Extracting ../data/MNIST/raw/train-labels-idx1-ubyte.gz to ../data/MNIST/raw



100%|██████████| 1648877/1648877 [00:00<00:00, 25573589.55it/s]


Extracting ../data/MNIST/raw/t10k-images-idx3-ubyte.gz to ../data/MNIST/raw



100%|██████████| 4542/4542 [00:00<00:00, 7650814.77it/s]


Extracting ../data/MNIST/raw/t10k-labels-idx1-ubyte.gz to ../data/MNIST/raw



In [6]:
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1,128,3, padding=1)
        self.bn1 = nn.BatchNorm2d(128)
        self.tns1 = nn.Conv2d(128,8,1, padding=1)
        self.conv2 = nn.Conv2d(8,16,3, padding=1)
        self.bn2 = nn.BatchNorm2d(16)
        self.pool1 = nn.MaxPool2d(2, 2)
        self.conv3 = nn.Conv2d(16,16,3, padding=1)
        self.bn3 = nn.BatchNorm2d(16)
        self.conv4 = nn.Conv2d(16,32,3, padding=1)
        self.bn4 = nn.BatchNorm2d(32)
       
        self.pool2 = nn.MaxPool2d(2, 2)
        self.tns2 = nn.Conv2d(32,16,1, padding=1)
        
        self.conv5 = nn.Conv2d(16,16,3, padding=1)
        self.bn5 = nn.BatchNorm2d(16)
        self.conv6 = nn.Conv2d(16,32,3, padding=1)
        self.bn6 = nn.BatchNorm2d(32)
        self.conv7 = nn.Conv2d(32,10,1, padding=1)
        self.gpool = nn.AvgPool2d(kernel_size=7)
        self.drop = nn.Dropout2d(0.1)
    def forward(self, x):
        x = self.tns1(self.drop(self.bn1(F.relu(self.conv1(x)))))
        x = self.drop(self.bn2(F.relu(self.conv2(x))))
        x = self.pool1(x)
        x = self.drop(self.bn3(F.relu(self.conv3(x))))
        x = self.drop(self.bn4(F.relu(self.conv4(x))))
        x = self.tns2(self.pool2(x))
        x = self.drop(self.bn5(F.relu(self.conv5(x))))
        x = self.drop(self.bn6(F.relu(self.conv6(x))))
        x = self.conv7(x)
        x = self.gpool(x)
        x = x.view(-1, 10)
        return F.log_softmax(x)

def model_summary(model,input_size):
        # use_cuda = torch.cuda.is_available()
        # device = torch.device("cuda" if use_cuda else "cpu")
        summary(model, input_size=input_size)

def GetCorrectPredCount(pPrediction, pLabels):
  return pPrediction.argmax(dim=1).eq(pLabels).sum().item()

def model_train(model, device, train_loader, optimizer, epoch):
    model.train()
    pbar = tqdm(train_loader)
    correct = 0
    processed = 0
    for batch_idx, (data, target) in enumerate(pbar):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        correct += GetCorrectPredCount(output, target)
        processed += len(data)
        pbar.set_description(desc= f'Train: Loss={loss.item():0.4f} Batch_id={batch_idx} Accuracy={100*correct/processed:0.2f}')


def model_test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)

    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.2f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))

In [7]:
model = Net().to(device) #setting model to switch to gpu
model_summary(model,(1,28,28)) #computing param size w.r.t input size of 28 x 28

----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1          [-1, 128, 28, 28]           1,280
       BatchNorm2d-2          [-1, 128, 28, 28]             256
         Dropout2d-3          [-1, 128, 28, 28]               0
            Conv2d-4            [-1, 8, 30, 30]           1,032
            Conv2d-5           [-1, 16, 30, 30]           1,168
       BatchNorm2d-6           [-1, 16, 30, 30]              32
         Dropout2d-7           [-1, 16, 30, 30]               0
         MaxPool2d-8           [-1, 16, 15, 15]               0
            Conv2d-9           [-1, 16, 15, 15]           2,320
      BatchNorm2d-10           [-1, 16, 15, 15]              32
        Dropout2d-11           [-1, 16, 15, 15]               0
           Conv2d-12           [-1, 32, 15, 15]           4,640
      BatchNorm2d-13           [-1, 32, 15, 15]              64
        Dropout2d-14           [-1, 32,

<ipython-input-6-28a0bca497bc>:37: UserWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.
  return F.log_softmax(x)


### Num Parameters in Model : 18,738

In [8]:
#Initializing optimizer
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.9)

#Running n_epochs
n_epochs = 21
for epoch in range(1, n_epochs):
    print(f"epoch {epoch}")
    model_train(model, device, train_loader, optimizer, epoch)
    model_test(model, device, test_loader)

epoch 1


  0%|          | 0/469 [00:00<?, ?it/s]<ipython-input-6-28a0bca497bc>:37: UserWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.
  return F.log_softmax(x)
Train: Loss=0.2443 Batch_id=468 Accuracy=77.72: 100%|██████████| 469/469 [00:21<00:00, 21.51it/s]



Test set: Average loss: 0.1285, Accuracy: 9663/10000 (96.63%)

epoch 2


Train: Loss=0.1369 Batch_id=468 Accuracy=95.50: 100%|██████████| 469/469 [00:22<00:00, 20.72it/s]



Test set: Average loss: 0.0605, Accuracy: 9826/10000 (98.26%)

epoch 3


Train: Loss=0.1198 Batch_id=468 Accuracy=96.92: 100%|██████████| 469/469 [00:21<00:00, 21.90it/s]



Test set: Average loss: 0.0478, Accuracy: 9857/10000 (98.57%)

epoch 4


Train: Loss=0.0666 Batch_id=468 Accuracy=97.56: 100%|██████████| 469/469 [00:21<00:00, 21.80it/s]



Test set: Average loss: 0.0360, Accuracy: 9880/10000 (98.80%)

epoch 5


Train: Loss=0.0501 Batch_id=468 Accuracy=97.87: 100%|██████████| 469/469 [00:21<00:00, 21.48it/s]



Test set: Average loss: 0.0337, Accuracy: 9896/10000 (98.96%)

epoch 6


Train: Loss=0.0221 Batch_id=468 Accuracy=98.02: 100%|██████████| 469/469 [00:22<00:00, 21.20it/s]



Test set: Average loss: 0.0308, Accuracy: 9904/10000 (99.04%)

epoch 7


Train: Loss=0.0707 Batch_id=468 Accuracy=98.27: 100%|██████████| 469/469 [00:21<00:00, 21.34it/s]



Test set: Average loss: 0.0246, Accuracy: 9919/10000 (99.19%)

epoch 8


Train: Loss=0.0167 Batch_id=468 Accuracy=98.44: 100%|██████████| 469/469 [00:22<00:00, 21.18it/s]



Test set: Average loss: 0.0233, Accuracy: 9915/10000 (99.15%)

epoch 9


Train: Loss=0.0581 Batch_id=468 Accuracy=98.53: 100%|██████████| 469/469 [00:21<00:00, 21.33it/s]



Test set: Average loss: 0.0227, Accuracy: 9919/10000 (99.19%)

epoch 10


Train: Loss=0.0629 Batch_id=468 Accuracy=98.55: 100%|██████████| 469/469 [00:21<00:00, 21.37it/s]



Test set: Average loss: 0.0240, Accuracy: 9921/10000 (99.21%)

epoch 11


Train: Loss=0.0266 Batch_id=468 Accuracy=98.69: 100%|██████████| 469/469 [00:21<00:00, 21.51it/s]



Test set: Average loss: 0.0221, Accuracy: 9923/10000 (99.23%)

epoch 12


Train: Loss=0.0276 Batch_id=468 Accuracy=98.67: 100%|██████████| 469/469 [00:20<00:00, 22.45it/s]



Test set: Average loss: 0.0210, Accuracy: 9922/10000 (99.22%)

epoch 13


Train: Loss=0.0221 Batch_id=468 Accuracy=98.75: 100%|██████████| 469/469 [00:20<00:00, 22.75it/s]



Test set: Average loss: 0.0209, Accuracy: 9928/10000 (99.28%)

epoch 14


Train: Loss=0.0742 Batch_id=468 Accuracy=98.76: 100%|██████████| 469/469 [00:21<00:00, 22.00it/s]



Test set: Average loss: 0.0208, Accuracy: 9920/10000 (99.20%)

epoch 15


Train: Loss=0.0865 Batch_id=468 Accuracy=98.80: 100%|██████████| 469/469 [00:20<00:00, 22.40it/s]



Test set: Average loss: 0.0178, Accuracy: 9935/10000 (99.35%)

epoch 16


Train: Loss=0.0582 Batch_id=468 Accuracy=98.86: 100%|██████████| 469/469 [00:21<00:00, 22.10it/s]



Test set: Average loss: 0.0185, Accuracy: 9937/10000 (99.37%)

epoch 17


Train: Loss=0.0105 Batch_id=468 Accuracy=98.87: 100%|██████████| 469/469 [00:21<00:00, 22.03it/s]



Test set: Average loss: 0.0222, Accuracy: 9925/10000 (99.25%)

epoch 18


Train: Loss=0.0631 Batch_id=468 Accuracy=98.95: 100%|██████████| 469/469 [00:21<00:00, 22.31it/s]



Test set: Average loss: 0.0203, Accuracy: 9939/10000 (99.39%)

epoch 19


Train: Loss=0.0751 Batch_id=468 Accuracy=98.95: 100%|██████████| 469/469 [00:20<00:00, 22.71it/s]



Test set: Average loss: 0.0202, Accuracy: 9929/10000 (99.29%)

epoch 20


Train: Loss=0.0367 Batch_id=468 Accuracy=99.00: 100%|██████████| 469/469 [00:21<00:00, 21.83it/s]



Test set: Average loss: 0.0159, Accuracy: 9942/10000 (99.42%)



# Reached Accuracy of 99.42 on Test Set in 20th epoch
